File: Olympiad/COI/official/2007/patrik.pas
(*
  Croatian Olympiad in Informatics 2007
  Task PATRIK
*)

const MAX = 500000;
var n, i, height, top : longint;
    stack : array[1..MAX] of longint;
    howmany : array[0..MAX] of longint;
    solution : int64;    
begin
    readln(n);
    for i:=1 to n do
    begin        
        readln(height);
        while (top>0)and(stack[top]<height) do 
        begin
            solution := solution + howmany[top];
            dec(top);
        end;
        if (top>0)and(stack[top]=height) then
        begin
            solution := solution + howmany[top];
            inc(howmany[top]);
            if top>1 then inc(solution);
        end else
        begin
            if top>0 then inc(solution);
            inc(top);
            howmany[top]:=1;            
            stack[top]:=height;
        end;
    end;    
    writeln(solution);
end.

File: Olympiad/COI/official/2007/patrik.cpp
/*
  Croatian Olympiad in Informatics 2007
  Task PATRIK
*/

#include <algorithm>
#include <cstdio>
#include <iostream>
#include <stack>

using namespace std;

typedef long long llint;
typedef pair<int,int> par;

stack<par> S;

int main( void ) {
   int n;
   scanf( "%d", &n );

   llint ret = 0;
   for( int i = 0; i < n; ++i ) {
      int h;
      scanf( "%d", &h );
      
      par p( h, 1 );
      for( ; !S.empty() && S.top().first <= h; S.pop() ) {
         ret += S.top().second;
         if( S.top().first == h ) p.second += S.top().second;
      }

      if( !S.empty() ) ++ret;
      S.push( p );
   }
   
   cout << ret << endl;

   return 0;
}
